lighthouse-logger/index.js

```diff
@@ -118,33 +174,149 @@ class Log {
     return level_ === 'verbose';
   }
 
-  static time({msg, id, args = []}, level = 'log') {
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
+  static time({msg, id, args}, level = 'log') {
     marky.mark(id);
-    Log[level]('status', msg, ...args);
+    Log[level]('status', msg, ...(args || []));
   }
 
-  static timeEnd({msg, id, args = []}, level = 'verbose') {
-    Log[level]('statusEnd', msg, ...args);
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
+  static timeEnd({msg, id, args}, level = 'verbose') {
+    Log[level]('statusEnd', msg, ...(args || []));
     marky.stop(id);
   }
 
+  /* eslint-disable no-invalid-this */
+  /**
+   * Decorates a function, calling time/timeEnd before/after calling the original function.
+   * @template T
+   * @template {*} R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeMsg = (_this, args) => {
+      if (typeof opts.msg === 'string') return opts.msg;
+      if (typeof opts.msg === 'function') return opts.msg.apply(_this, args);
+      throw new Error('expected msg');
+    };
+
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeId = (_this, args) => {
+      if (typeof opts.id === 'string') return opts.id;
+      if (typeof opts.id === 'function') return opts.id.apply(_this, args);
+      return `lh:${originalFn.name}`;
+    };
+
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      const status = {msg: computeMsg(this, args), id: computeId(this, args)};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.apply(this, args);
+      } catch (err) {
+        Log.timeEnd(status, timeEndLogLevel);
+        // intercept any errors and elide the time decoration from the stack trace
```

Comment:
Sooooo...I should have been more clear in https://github.com/GoogleChrome/lighthouse/issues/6425#issuecomment-434120968. I really meant, *if* the stack traces are bad we should consider this. Are they? Because if not it can be really confusing why all the function calls that actually happened aren't in the trace.

Maybe we should punt on this and add it later if we find the stack traces difficult to work with? Or have you found it helpful to get clean stack traces so far?

(I'm also curious how official decorators plan on handling this)

Reply:
It didn't really come up, so I can't say I found it useful.

Monitoring tools like Rollbar do their best to clean up calls to their lib functions from stack traces. But our case is very minor, and theirs is an library meant for monitoring ...

I could go either way with this. If you think of a stack trace as primarily a tool developers use to navigate code flow, It's a nice idea to hide function calls. If you want to see exactly the whole picture, you wouldn't want things hidden. I imagine it won't ever come up again either way we go with.